codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java

```diff
@@ -526,32 +529,29 @@ private boolean exceedsMaxHeaderSize(long size) {
         return true;
     }
 
-    private byte[] readStringLiteral(InputStream in, int length) throws IOException {
-        byte[] buf = new byte[length];
-        if (in.read(buf) != length) {
-            throw DECOMPRESSION_EXCEPTION;
-        }
-
+    private CharSequence readStringLiteral(ByteBuf in, int length) throws IOException {
         if (huffmanEncoded) {
-            return Huffman.DECODER.decode(buf);
+            return huffmanDecoder.decode(in, length);
         } else {
-            return buf;
+            byte[] buf = new byte[length];
+            in.readBytes(buf);
+            return new AsciiString(buf, false);
```

Comment:
AsciiString is a bit overloaded. IIUC it was originally used to provide memory benefit over using a collection backed by `char[]`, and only supported encodings where the `char` value could be directly represented as a `byte`. However it can be used more as a "byte string" (which can be modified via direct access to the underlying `byte[]`). `AsciiString` and `ByteString` were at one point separate classes ... but then were combined to reduce complexity (and awkward dependencies between the two). Now `AsciiString` even supports methods which take a `Charset` and only defaults to the "`char` which can be represented as a `byte`" behavior if no `Charset` or `byte[]` is provided.